src/utils/use-search-params-state.ts

```diff
@@ -0,0 +1,48 @@
+import { useCallback, useEffect, useState } from 'react';
+import type { NavigateOptions } from 'react-router';
+import { useNavigation, useSearchParams } from '@remix-run/react';
+import { mergeUrlSearchParams } from './index';
+
+export interface SearchParamsStateConverter<S> {
+    fromSearchParams: (params: URLSearchParams) => S;
+    toSearchParams: (state: S) => URLSearchParams;
+}
+
+/**
+ * Helper to manage the state that lives in URL search params.
+ * Search params are not updated immediately
```

Comment:
Not updated immediately by what? By the hook?

Reply:
By Remix

Reply:
It's not clear from the comment.

Also the following "because": I'm not sure how one leads to another. Remix could `history.pushState` immediately, it doesn't do that, but not because of navigation.